### PR TITLE
[hail] Correctly merge files in ExportPlink

### LINKED ISSUE #4932: copyMerge is broken - needs to take a list of things to merge
https://hail.zulipchat.com/#narrow/stream/123010-Hail-0.2E2.20support/subject/export_plink.20error

### PR DESCRIPTION
**Pull request by danking**
### Problem Description

`ExportPlink` was previously modified to resiliently handle failure of a
Spark task by including a per-task UUID. `copyMerge` was not modified to
correctly handle the directories generated by this modified
`ExportPlink`.

For example, if exactly one task out of N fails during `ExportPlink`, the
temporary output directory will contain N+1 partition files. One of
these partition files is corrupted and invalid. The other N are the
output of successful task completion. The invalid file should simply be
ignored by `copyMerge`.

### Changes Made

This PR modifies `copyMerge` to take an optional list of files to
merge. If that argument is set to `None`, the original behavior
persists. The original behavior is used by `RichRDD.writeTable` and
`RichRDDByteArray.saveFromByteArrays`. These two methods use the default
Spark parallel export system. This system is not resilient to all task
failures, but *does* ensure failed tasks do not generate garbage
partitions in the output directory. Ergo, they can safely use the
original behavior of `copyMerge`.

### On Testing

I do not test this behavior because failing a task during write is a
little bit tricky. I have verified that all users of `copyMerge` now use
`copyMerge` correctly. Adding a test to `ExportPlink` would not save us
from incorrectly using `copyMerge` in the future.

A longer term testing strategy that includes a Chaos Monkey that kills
entire containers during Hail Scala tests execution would protect
against this type of bug.

---

Fixes #4932

### CODE REVIEW

File: hail/src/main/scala/is/hail/io/plink/ExportPlink.scala [hunk 1]
```diff
@@ -104,15 +104,17 @@ object ExportPlink {
         }
       }
 
-      Iterator.single(rowCount)
-    }.collect().sum
+      Iterator.single(f -> rowCount)
+    }.collect().foldLeft(
+      (FastIndexedSeq[String](), 0L)
+    ) { case ((files, sum), (f, n)) => (files :+ f, sum + n) }
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/io/plink/ExportPlink.scala]: This is potentially quadratic

Comment (danking) [reply hail/src/main/scala/is/hail/io/plink/ExportPlink.scala]: Is `FastIndexedSeq` really that dumb?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/io/plink/ExportPlink.scala]:
```
  def :+[B >: A, That](elem: B)(implicit bf: CanBuildFrom[Repr, B, That]): That = {
    val b = bf(repr)
    b ++= thisCollection
    b += elem
    b.result()
  }
```

Comment (danking) [reply hail/src/main/scala/is/hail/io/plink/ExportPlink.scala]: oh dear jesus the implementation of WrappedArray.canBuildFrom makes me want to cry.

**Comment by danking** [comment]
That `:+` is anything other than exponential widening of the array is an offense to all of CS.

File: hail/src/main/scala/is/hail/io/plink/ExportPlink.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/io/plink/ExportPlink.scala]: image

**Comment by tpoterba** [comment]
pushed a fix

**Pull request merged**